spaceship/lib/spaceship/tunes/tunes_client.rb

```diff
@@ -575,12 +575,25 @@ def candidate_builds(app_id, version_id)
     #####################################################
 
     # @param (testing_type) internal or external
-    def build_trains(app_id, testing_type, platform: nil)
+    def build_trains(app_id, testing_type, platform: nil, tries: 5)
       raise "app_id is required" unless app_id
       url = "ra/apps/#{app_id}/trains/?testingType=#{testing_type}"
       url += "&platform=#{platform}" unless platform.nil?
       r = request(:get, url)
-      parse_response(r, 'data')
+      return parse_response(r, 'data')
+    rescue Spaceship::Client::UnexpectedResponse => ex
+      # Build trains fail randomly very often
+      # we need to catch those errors and retry
+      # https://github.com/fastlane/fastlane/issues/6419
+      error_content = ex.to_s
+      if error_content == "{\"data\"=>nil, \"messages\"=>{\"warn\"=>nil, \"error\"=>[\"ITC.response.error.OPERATION_FAILED\"], \"info\"=>nil}, \"statusCode\"=>\"ERROR\"}"
+        unless (tries -= 1).zero?
```

Comment:
This is a personal preference, (and somewhat nitpicky) but I strongly dislike conditional checks that include a side-effect.

The reason is that it becomes possible to accidentally prevent the side-effect from happening in later edits, by introducing a short-circuiting operation to the conditional. So, later on someone edits it to be:

`unless other_thing && (tries -= 1).zero?`

and if `other_thing` causes a short-circuited evaluation, now your counting is possibly wrong, or infinitely looping.

So my preference is to do the side-effect operation separately, then do the boolean expression on it. :+1: